pkg/util/filewatch.go

```diff
@@ -0,0 +1,98 @@
+package util
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/pkg/flow/logging"
+)
+
+type ReloadFunc func() error
+
+type FileWatcher struct {
+	filePath     string
+	reloadFunc   ReloadFunc
+	pollInterval time.Duration
+	forcePolling bool
+}
+
+// NewFileWatcher creates a new file watcher that will call reloadFunc when the file at filePath is modified.
+func NewFileWatcher(filePath string, reloadFunc ReloadFunc, pollInterval time.Duration, forcePolling bool) *FileWatcher {
+	return &FileWatcher{
+		filePath:     filePath,
+		reloadFunc:   reloadFunc,
+		pollInterval: pollInterval,
+		forcePolling: forcePolling,
+	}
+}
+
+// Watch starts watching the file for changes and calls reloadFunc when the file is modified.
+func (fw *FileWatcher) Watch(l *logging.Logger, ctx context.Context, wg *sync.WaitGroup) {
+	wg.Add(1)
+	go fw.watch(l, ctx, wg)
+}
+
+func (fw *FileWatcher) watch(l *logging.Logger, ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		level.Error(l).Log("msg", "failed to create watcher", "err", err)
+		return
+	}
+	defer watcher.Close()
+
+	err = watcher.Add(fw.filePath)
+	if err != nil {
+		level.Error(l).Log("msg", "failed to add file to watcher", "err", err)
+		return
+	}
+
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
+	defer signal.Stop(sigCh)
```

Comment:
These signals are already handled by the caller so you don't need to watch for them here; instead you can just wait until the provided context gets canceled to exit.